### ISSUE TITLE
Steps always refer first test that was run

### ISSUE THREAD

**Issue by ilyapuchka** [net-a-porter-mobile/XCTest-Gherkin#104]
Steps definers are loaded only once, when first test is started, so they will keep the reference to it. When next step starts this reference is not updated, so if steps definers try to use it to access any property of currently running test case they will actually get values from previous tests. We are using `launchArguments` property in our XCTestCase subclasses to configure the app and some steps are designed to change these arguments. So when several tests are running with the same steps, they will all change first test property instead of the currently running test.

**Comment by deanWombourne** [comment]
Hey. I'm hesitant to redefine all the steps each test run because I don't know the cost of doing that (it would depend on what all the other users o this library are doing in their StepDefiners!) - but it does seem like tying all the step definitions to one initial test seems the wrong thing to do!

Can you post a minimal example of the issue you're facing - that way I can have a play around and make sure I'm solving your problem. Would also help when writing a test for it ;)

**Comment by ilyapuchka** [comment]
It's not the problem of re-registreing all steps, just updating the reference to current test. Probably you can utilise `XCTestObservationCenter` for that.
Here is a basic example: 

```swift
class BaseFeature: XCTestCase {
  var launchArguments: [String] = []

  func setUp() {
    launchArguments += ["-uitests", "YES"]
    app.launchArguments = launchArguments
    app.launch()
  }

  func tearDown() {
    launchArguments = []
    app.launchArguments = []
  }

  func testFirst() {
     When("app is launched with some parameter")
     ...
  }

  func testSecond() {
     When("app is launched with some parameter")
     ...
  }
}

class LoginSteps: StepsDefiner {
  var feature: BaseFeature {
    return test as! BaseFeature
  }

  func defineSteps() {
    step("app is launched with some parameter") {
       feature.launchArguments += ["some parameter"]
    }
  }
}
```

When both tests are run the second test will fail because app will not get all launch arguments. When `testSecond` is calling `"app is launched with some parameter"` step, the step still has a reference to `testFirst` which has it's `launchArguments` cleared after `tearDown`. So step and test basically operate on different `launchArgument` properties.

**Comment by deanWombourne** [comment]
OK, I had trouble getting your example to reproduce the bug (don't ask, had a fight with XCUIApplicatoin().launchArguments! ) but think this reproduces it as well:

```swift
import Foundation
import XCTest
import XCTest_Gherkin

final class BaseFeature: XCTestCase {
    var launchArguments: [String] = []

    override func setUp() {
        launchArguments += ["-uitests", "YES"]
    }

    override func tearDown() {
        launchArguments = []
    }

    func testFirst() {
        When("app is launched with some parameter")
        Then("launch parameters should contain 3 values")
    }

    func testSecond() {
        When("app is launched with some parameter")
        Then("launch parameters should contain 3 values")
    }
}

final class LoginSteps: StepDefiner {

    var feature: BaseFeature {
        return test as! BaseFeature
    }

    override func defineSteps() {
        step("app is launched with some parameter") {
            self.feature.launchArguments += ["some parameter"]
        }

        // 2 from the test + the 'some parameter'
        step("launch parameters should contain 3 values") {
            XCTAssertEqual(self.feature.launchArguments.count, 3)
        }
    }
}
```

Also, that's now a test case in the library, so I have something to work against :) `testFirst` passes, and `testSecond` fails, due to the bug you've found where test second is still looking at the first test instance :| 

Let me know if this is a valid test, and I'll work on getting it fixed :)

**Comment by deanWombourne** [comment]
^ Excuse the edit - I've narrowed it down into a minimal test case (which can now be ran as a unit test so that make fixing it much faster ;)

**Comment by deanWombourne** [comment]
Sorry, pressed the wrong button - definitely didn't mean to close it!

**Comment by deanWombourne** [comment]
FYI: I've just released this as version 0.14.1 so a pod update should fix it :)
